### PR TITLE
Wrong error message in CSV importer

### LINKED ISSUE #46505: Built-in product importer error message does not differentiate between SKU and ID
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

This is a bit pedantic I know, but details make a difference.

When importing a product whose SKU already exists on the site, the importer fails to import the product and shares an error, this is expected.

The error says `A product with this ID already exists.` this is inaccurate and confusing, given that reason for the error is because the SKU exists on another product.

### Expected behavior

Error message should read `A product with this SKU already exists.` when the reason for the import failure is because a matching SKU already exists. 

### Actual behavior

The error conflates SKU and ID which is not accurate and confusing. To be fair, the ID shown for the product **is** the ID of the existing product (albeit not the one in the CSV)
import error confounding

### Steps to reproduce

1. On a brand new test site, import the sample_products.csv that is included in the plugin files at `/woocommerce/sample-data/`
2. Import the same CSV a second time.
3. Note the error. ID doesn't match spreadsheet, SKU does. 

### WordPress Environment

<details>

```
### WordPress Environment ###

WordPress address (URL): http://woo-docs.local
Site address (URL): http://woo-docs.local
WC Version: 8.7.0
REST API Version: ✔ 8.7.0
Action Scheduler Version: ✔ 3.7.2
Log Directory Writable: ✔
WP Version: 6.5.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.1.23
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.4.0
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.7.0
WC Database Prefix: wp_
Total Database Size: 7.07MB
Database Data Size: 5.35MB
Database Index Size: 1.72MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.09MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 4.06MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.09MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 27
page: 7
post: 2
product: 37
product_variation: 14
wp_font_family: 12
wp_navigation: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (2) ###

WooCommerce: by Automattic – 8.7.0
WordPress Importer: by wordpressdotorg – 0.8.2

### Inactive Plugins (0) ###


### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 13 B

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.1
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /Users/briantaylor/Local Sites/woo-docs/app/public/wp-content/plugins/woocommerce/templates/block-notices/error.php
/Users/briantaylor/Local Sites/woo-docs/app/public/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/Users/briantaylor/Local Sites/woo-docs/app/public/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-04-12 00:33:39 +00:00
Options: ✔
Notes: 62
Onboarding: completed

### Action Scheduler ###

Complete: 319
Oldest: 2024-04-11 00:41:26 +0000
Newest: 2024-04-11 20:02:19 +0000

Failed: 2
Oldest: 2024-04-11 19:20:49 +0000
Newest: 2024-04-11 19:20:49 +0000

Pending: 1
Oldest: 2024-04-12 00:35:01 +0000
Newest: 2024-04-12 00:35:01 +0000


### Status report information ###

Generated at: 2024-04-11 20:11:03 +00:00
```

</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by websupporter**
### Changes proposed in this Pull Request:

If you import a CSV and opt for not updating existing products and you try to import a product whose SKU already exists, you get the message that the product was not imported because the ID already existed. This PR fixes that behavior. It proposes to first check for the SKU before checking for the ID.

The reason why we can end up in the "ID already exists" message is that the ID we have at hand at this point is already normalized to the product ID in our DB, which we retrieve from the SKU and no longer the ID from the CSV. By simply switching the two checks, we show the correct error.

Closes #46505 .

### How to test the changes in this Pull Request:

1. Create a WooCommerce shop with one product
2. Export the product (all fields)
3. Change the ID in the exported file to an ID that does not exist
4. Import the file, but instruct the importer to not update existing products
5. You should see 1 skipped product
6. Open the log, the reason stated for skipping should be: "A product with this SKU already exists." It should no longer be "A product with this ID already exists."

### CODE REVIEW

**Review by barryhughes — APPROVED** [review]
Given the described scenario, I now see the more appropriate (SKU-related) error.

<div align="center">Product importer results, showing 'existing SKU' error.</div>

Thanks also for adding the test. We'll wait on CI passing, then should be able to proceed!

**Comment by barryhughes** [comment]
> **Core e2e tests 1/6 - @woocommerce/plugin-woocommerce [e2e]**
> `activate-and-setup/core-profiler.spec.js:184:3` › Store owner can complete the core profiler › Can complete the core profiler installing default extensions

Same test continues to fail. I don't see the connection with this PR, though, so will give it another spin.

File: plugins/woocommerce/includes/import/class-wc-product-csv-importer.php [hunk 1]
```diff
@@ -1135,24 +1135,24 @@ public function import() {
...
 				);
 				continue;
 			}
 
-			if ( $sku_exists && ! $update_existing ) {
+			if ( $id_exists && ! $update_existing ) {
 				$data['skipped'][] = new WP_Error(
 					'woocommerce_product_importer_error',
-					esc_html__( 'A product with this SKU already exists.', 'woocommerce' ),
+					$id . esc_html__( 'A product with this ID already exists.', 'woocommerce' ),
```
Comment (barryhughes) [inline plugins/woocommerce/includes/import/class-wc-product-csv-importer.php]:
```suggestion
					esc_html__( 'A product with this ID already exists.', 'woocommerce' ),
```

**Comment by websupporter** [comment]
Thank you @barryhughes

**Comment by barryhughes** [comment]
Unsure why the linter is failing (it's failing to run, it seems, rather than detecting a problem) 🤔

**Comment by barryhughes** [comment]
It seems to be a bad day for running CI checks 🤦🏼

**Pull request merged**